File: FindPlacesNearMe/FindPlacesNearMe/ViewController.m
//
//  ViewController.m
//  FindPlacesNearMe
//
//  Copyright © 2020 Amor Rio. All rights reserved.
//

#import "ViewController.h"
#import "CoreDataHelper.h"
#import "Place+CoreDataProperties.h"
#import "AppDelegate.h"

@interface ViewController ()
//views
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *searchingIndicator;
@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
@property (weak, nonatomic) IBOutlet UIButton *resetButton;
@property (weak, nonatomic) IBOutlet UIButton *refreshButton;

@property (strong, nonatomic) CLLocationManager *locationManager;
@property (strong, nonatomic) NSArray *currentAnnotations;
@property (strong, nonatomic) NSString *currentSearchString;
@property (strong, nonatomic) CoreDataHelper *dbHelper;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if ([CLLocationManager locationServicesEnabled])
    {
        self.locationManager = [[CLLocationManager alloc] init];
        self.locationManager.delegate = self;
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBest;
        [self.locationManager requestWhenInUseAuthorization];
        
        if ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways ||
            [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedWhenInUse)
        {
            self.mapView.showsUserLocation = YES;
        }
    }
    
    AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    self.dbHelper = [[CoreDataHelper alloc] initWithManagedObjectContext:appDelegate.persistentContainer.viewContext];
}

- (IBAction)refreshButtonHit:(id)sender {
    [self performSearch];
}

- (IBAction)resetButonHit:(id)sender {
    if ([self.currentAnnotations count] > 0){
        
        for (id<MKAnnotation> annotation in self.mapView.selectedAnnotations)
        {
            [self.mapView deselectAnnotation:annotation animated:NO];
        }
        
        [self.mapView showAnnotations:self.currentAnnotations animated:YES];
    }

}
#pragma mark MKMapViewDelegate Methods

- (MKAnnotationView *)mapView:(MKMapView *)mapView 
            viewForAnnotation:(id<MKAnnotation>)annotation
{
    if ([annotation isKindOfClass:[MKUserLocation class]])
    {
        //use system view for user's current location
        return nil;
    }
    
    NSString *identifier = @"MyAnnotation";
    MKAnnotationView *annotationView = [mapView dequeueReusableAnnotationViewWithIdentifier:identifier];
    
    if (annotationView == nil)
    {
        annotationView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:identifier];
        annotationView.canShowCallout = YES;
    }
    else
    {
        [annotationView setAnnotation:annotation];
    }
    
    return annotationView;
}

- (void)mapView:(MKMapView *)mapView didAddAnnotationViews:(NSArray<MKAnnotationView *> *)views
{
    //when the annotations are already shown in map, stop the searching indicator
    //and enable user interaction
    [self didEndSearch];
    self.resetButton.enabled = YES;
}


#pragma mark - CLLocationManagerDelegate Methods

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    if ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways ||
        [CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedWhenInUse)
    {
        self.mapView.showsUserLocation = YES;
    }
}


#pragma mark - UISearchBarDelegate Methods

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    //hide the keyboard
    [searchBar resignFirstResponder];

    //remove leading and trailing white spaces
    NSString *textToSearch = [searchBar.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    
    if (textToSearch == nil || [textToSearch length] == 0)
    {
        self.currentSearchString = nil;
        return;
    }

    self.currentSearchString = textToSearch;
    
    [self performSearch];
    

}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    //if the search is cancelled, hide the keyboard
    [searchBar resignFirstResponder];
}

#pragma mark - Utility Methods

-(void)addAnnotationsToMapView:(NSArray *)annotations
{
    if (annotations == nil || [annotations count]  == 0)
    {
        [self showErrorSearchAlertWithMessage:@"Place with name is not found within the area"];
        [self didEndSearch];
        return;
    }
    
    self.currentAnnotations = annotations;
    [self.mapView addAnnotations:annotations];
    //this adjusts the view to visibly show all annotations in the map
    [self.mapView showAnnotations:annotations animated:YES];
}

- (void)performSearchForSearchString:(NSString *) searchString
{
    NSURL *requestURL = [self createPlaceRestAPIRequestURLForSearchString:searchString coordinate:self.mapView.userLocation.coordinate];
    
    NSMutableURLRequest *request= [NSMutableURLRequest requestWithURL:requestURL];
    
    NSURLSessionDataTask *dataTask = [[NSURLSession sharedSession] dataTaskWithRequest:request
                                                                     completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if (httpResponse.statusCode == 200)
        {
            NSError *error;
            NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:&error];
            
            NSArray *annotations = [self annotationsForSearchResult:result];
            if (annotations != nil)
            {
                //perform in main queue
                [self performSelectorOnMainThread:@selector(addAnnotationsToMapView:) withObject:annotations waitUntilDone:NO];
            }
        }
        else {
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self showErrorSearchAlertWithMessage:@"Searching failed"];
                [self didEndSearch];
            });
        }
    }];
    
    [dataTask resume];
}

- (void)performSearch
{
    if (self.currentSearchString == nil)
    {
        return;
    }
    
    [self.mapView removeAnnotations:self.currentAnnotations];
    self.currentAnnotations = nil;
    NSArray *places = [self.dbHelper placesForSearchString:self.currentSearchString coordinate:self.mapView.userLocation.location.coordinate];
    if (places && [places count] > 0){
        self.currentAnnotations = [self annotationsForPlaces:places];
        [self addAnnotationsToMapView:self.currentAnnotations];
        return;
    }
    
    self.refreshButton.enabled = NO;
    self.resetButton.enabled = NO;
    //show activity indicator while searching
    [self.searchingIndicator startAnimating];
    //disable interaction with the search bar and the map
    self.mapView.userInteractionEnabled = NO;
    self.searchBar.userInteractionEnabled = NO;
    
    [self performSearchForSearchString:self.currentSearchString];
}

- (NSURL *)createPlaceRestAPIRequestURLForSearchString:(NSString *)searchString coordinate:(CLLocationCoordinate2D)coordinate
{
    //build request URL using NSURLComponents
    //initialize with the Places Search Rest API URL
    NSURLComponents *restAPIURLComponents = [NSURLComponents componentsWithString:@"https://places.demo.api.here.com/places/v1/discover/search"];
    //add search parameters
    NSURLQueryItem *searchItem = [[NSURLQueryItem alloc] initWithName:@"q" value:searchString];
    //maximum of 10 results
    NSURLQueryItem *searchSize = [[NSURLQueryItem alloc] initWithName:@"size" value:@"10"];
    //search area within 5km around the current location
    NSString *areaSpecification = [NSString stringWithFormat:@"%lf,%lf;r=5000", coordinate.latitude, coordinate.longitude];
    NSURLQueryItem *searchArea = [[NSURLQueryItem alloc] initWithName:@"in" value:areaSpecification];
    
    //authentication credentials for demo account
    NSURLQueryItem *app_id = [[NSURLQueryItem alloc] initWithName:@"app_id" value:@"DemoAppId01082013GAL"];
    NSURLQueryItem *app_code = [[NSURLQueryItem alloc] initWithName:@"app_code" value:@"AJKnXv84fjrb0KIHawS0Tg"];
    
    //add the parameters
    NSArray *queryItems = [NSArray arrayWithObjects:searchItem, searchSize, searchArea, app_id, app_code, nil];
    [restAPIURLComponents setQueryItems:queryItems];
    
    return [restAPIURLComponents URL];
}

-(NSArray *)annotationsForPlaces:(NSArray *)places
{
    NSMutableArray *annotations = [[NSMutableArray alloc] init];
    
    for (Place *place in places){
         CLLocationCoordinate2D coordinate;
        coordinate.latitude = place.latitude;
        coordinate.longitude =place.longitude;
        
   
        NSString *distanceUnit = @"meters";
        double distanceDivisor = 1.0f;
        if ((place.distance - 1000.0f) > 0.0001f )
        {
            distanceDivisor = 1000;
            distanceUnit = @"km";
            
        }
        //set subtitle to distance description from location of user
        NSString *subTitle = [NSString stringWithFormat:@"%g %@ away", place.distance/distanceDivisor, distanceUnit];
        
        MKPointAnnotation *annotation = [[MKPointAnnotation alloc] init];
        annotation.coordinate = coordinate;
        //title and subtitle will show in callout when pins are tapped
        annotation.title = [NSString stringWithFormat:@"%@ %@", place.title, place.vicinity];
        annotation.subtitle = subTitle;
        
        
        [annotations addObject:annotation];
    }
    return annotations;
}

- (NSArray*)annotationsForSearchResult:(NSDictionary *)searchResult
{
    NSArray *places = [self.dbHelper placesForSearchResult:searchResult searchString:self.currentSearchString coordinate:self.mapView.userLocation.location.coordinate];
    
    return [self annotationsForPlaces:places];
}

-(void)showErrorSearchAlertWithMessage:(NSString *)errorMessage
{
    NSString *appName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleDisplayName"];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:appName message:errorMessage preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        
    }];
    
    [alertController addAction:action];
    
    [self presentViewController:alertController animated:YES completion:nil];
    
}

- (void)didEndSearch
{
    [self.searchingIndicator stopAnimating];
    self.refreshButton.enabled = YES;
    self.mapView.userInteractionEnabled = YES;
    self.searchBar.userInteractionEnabled = YES;
}

@end

File: FindPlacesNearMeSwift/FindPlacesNearMeSwift/ViewController.swift
//
//  ViewController.swift
//  FindPlacesNearMeSwift
//
//  Copyright © 2020 Amor Rio. All rights reserved.
//

import UIKit
import MapKit
import Foundation

class ViewController: UIViewController, MKMapViewDelegate, CLLocationManagerDelegate, UISearchBarDelegate
{
    @IBOutlet weak var mapView: MKMapView!
    @IBOutlet weak var searchBar: UISearchBar!
    @IBOutlet weak var searchingIndicator: UIActivityIndicatorView!
    @IBOutlet weak var refreshButton: UIButton!
    @IBOutlet weak var resetButton: UIButton!
    @IBOutlet weak var locationPermissionWarning: UILabel!

    let annotationViewIdentifier = "MyAnnotation"
    var locationManager : CLLocationManager?
    var currentAnnotations : [MKAnnotation]?
    var currentSearchString : String?
    var dbhelper : CoreDataHelper?

    override func viewDidLoad() {
        super.viewDidLoad()

        self.locationManager = CLLocationManager()
        self.locationManager?.delegate = self

        if CLLocationManager.locationServicesEnabled()
        {
            self.locationManager?.desiredAccuracy = kCLLocationAccuracyBest
            self.locationManager?.requestWhenInUseAuthorization()

            if CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedAlways ||
                CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedWhenInUse
            {
                //show user location only if location permission is granted
                self.mapView.showsUserLocation = true
            }
        }
        else {
            self.searchBar.isUserInteractionEnabled = false
            self.locationPermissionWarning.isHidden = false
        }

        self.mapView.register(MKPinAnnotationView.self, forAnnotationViewWithReuseIdentifier: self.annotationViewIdentifier)

        // Do any additional setup after loading the view.
        let delegate : AppDelegate  = UIApplication.shared.delegate as! AppDelegate
        self.dbhelper = CoreDataHelper(context: delegate.persistentContainer.viewContext)
    }


    ///The refreshButton reperforms the search given the search string that is currently in the search bar
    ///This is in case the user have changed location and wants to update the search result for user's new location
    @IBAction func refreshButtonHit(_ sender: Any) {
        self.performSearch()
    }

    ///The resetButton resets the map view to the initial view after a search
    ///That is the mapView is zoomed in such a way that all search result pins are visible in the map
    @IBAction func resetButtonHit(_ sender: Any) {
        if let annotations = self.currentAnnotations, annotations.count > 0 {

            for annotation in self.mapView.selectedAnnotations{
                self.mapView.deselectAnnotation(annotation, animated: false)
            }

            self.mapView.showAnnotations(annotations, animated: true)
        }
    }

    //MARK: - MKMapViewDelegate methods

    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {

        if annotation is MKUserLocation {
            return nil
        }

        var annotationView  = mapView.dequeueReusableAnnotationView(withIdentifier: self.annotationViewIdentifier)

        if annotationView == nil{
            annotationView = MKPinAnnotationView(annotation:annotation, reuseIdentifier:self.annotationViewIdentifier)
        }
        else {
            annotationView?.annotation = annotation
        }
        annotationView?.canShowCallout = true
        return annotationView
    }

    func mapView(_ mapView: MKMapView, didAdd views: [MKAnnotationView]) {
        //show end search state after the annotations are already shown in view
        self.didEndSearch()
        //enable only the reset button if there are annotations shown
        self.resetButton.isEnabled = true
    }

    //MARK: CLLocationManagerDelegate methods

    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        if CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedAlways ||
            CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedWhenInUse
        {
            self.mapView.showsUserLocation = true
            self.searchBar.isUserInteractionEnabled = true
            self.locationPermissionWarning.isHidden = true
            if self.currentSearchString != nil {
                self.refreshButton.isEnabled = true
            }

            if let annotations = self.currentAnnotations {
                self.mapView.addAnnotations(annotations)
                self.mapView .showAnnotations(annotations, animated: true)
                self.resetButton.isEnabled = true
            }
        }
        else {
            self.mapView.showsUserLocation = false
            self.searchBar.isUserInteractionEnabled = false
            self.refreshButton.isEnabled = false
            self.locationPermissionWarning.isHidden = false
            if let annotations = self.currentAnnotations {
                self.mapView.removeAnnotations(annotations)
                self.resetButton.isEnabled = false
            }
        }
    }

    //MARK: UISearchBarDelegate methods

    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
        let textToSearch = searchBar.text?.trimmingCharacters(in: NSCharacterSet.whitespacesAndNewlines)

        if textToSearch == nil || textToSearch?.count == 0
        {
            self.currentSearchString = nil;
            return
        }

        self.currentSearchString = textToSearch
        self.performSearch()

    }

    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }

    //MARK: UtilityMethods
    /**
    Adds array of annotations to the map view
    Sets the map to show all the annotations added

		-Parameter annotations: array of annotations to add to the mapView
    */
    func addAnnotationToMapView(annotations : [MKAnnotation]){
        self.currentAnnotations = annotations
        self.mapView.addAnnotations(annotations)
        self.mapView .showAnnotations(annotations, animated: true)
    }

    /**
    Performs the search for the current search string and show the result
    in the map
    It first tries to find if there is already a cached search result for the
    given search string in the current user location in the database.
    If there is no cached result, triggers the search using the places demo API
    */
    func performSearch() {
        if let searchString = self.currentSearchString {
            if let annotations = self.currentAnnotations {
                self.mapView.removeAnnotations(annotations)
                self.currentAnnotations = nil
            }

            //check first if search result for the given key and location is already saved in database
            //use saved result instead of querying in places demo api if there is saved result
            if let places = self.dbhelper?.placesForSearchResult(searchString: searchString, coordinate:self.mapView.userLocation.coordinate) {
                if let annotations =  self.annotationsForPlaces(places: places){
                    self.currentAnnotations = annotations
                    self.addAnnotationToMapView(annotations:annotations)
                }
            } else {
                //else perform a new search to the places demo api
                //disable the buttons
                self.refreshButton.isEnabled = false
                self.resetButton.isEnabled = false
                //start searching indicator
                self.searchingIndicator.startAnimating()
                //disable user interaction for mapView and searchBar
                self.mapView.isUserInteractionEnabled = false
                self.searchBar.isUserInteractionEnabled = false

                self.performSearchForString(searchString: searchString)
            }
        }
    }

    /**
    Performs the search using the places demo API  for the given searchString

    - Parameter searchString: place identifier to be searched
    around the current user location
    */
    func performSearchForString(searchString : String)
    {
        //get user coordinate
        let coordinate = self.mapView.userLocation.coordinate;

        //form the REST API url for the places demo API
        guard let url = self.placeRestAPIURLForSearchString(searchString: searchString, coordinate: coordinate)
            else {
              //if search url is not properly formed, show alert that search has failed
              //and end search
                self.showAlert(message:"Search failed")
                self.didEndSearch()
                return
        }

        //perform search via REST API using place demo API
        let dataTask = URLSession.shared.dataTask(with: url) { (data, response, error) in

            if  let httpresponse = response as? HTTPURLResponse {
                if httpresponse.statusCode == 200 {
                    do {
                        if let jsonData = data {
                            let result = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String:AnyObject]

                            if let annotations = self.annotationsForSearchResult(searchResult: result, searchString: searchString, coordinate: coordinate) {

                                DispatchQueue.main.async {
                                    if annotations.count == 0
                                    {
                                        self.showAlert(message:"Place with name is not found within the area" )
                                        self.didEndSearch()
                                        return
                                    }
                                    self.addAnnotationToMapView(annotations: annotations)
                                }
                                return
                            }
                        }
                    }
                    catch { }
                }
            }
            DispatchQueue.main.async {
                self.showAlert(message:"Search failed")
                self.didEndSearch()
            }
        }

        dataTask.resume()
    }

    /**
    Form the REST API URL using the places demo API  for the given searchString
    - Parameter searchString: place identifier to be searched
    around the current user location

      - Parameter coordinate: coordinates for the current user location
    */
    func placeRestAPIURLForSearchString(searchString : String, coordinate: CLLocationCoordinate2D) ->URL?
    {
        guard let restAPIURLComponents  = NSURLComponents(string:"https://places.demo.api.here.com/places/v1/discover/search") else {
            return nil
        }
        //add search parameters
        let searchItem = URLQueryItem(name:"q", value: searchString)

        //maximum of 10 results
        let searchSize = URLQueryItem(name:"size", value:"10")

        let lat = coordinate.latitude
        let lon = coordinate.longitude

        //search area within 5km around the current location
        let areaSpecification = "\(lat),\(lon);r=5000"

        let searchArea = URLQueryItem(name:"in", value:areaSpecification)

        //authentication credentials for demo account
        let app_id = URLQueryItem(name:"app_id", value:"DemoAppId01082013GAL")
        let app_code = URLQueryItem(name:"app_code", value:"AJKnXv84fjrb0KIHawS0Tg")

        restAPIURLComponents.queryItems = [searchItem, searchSize, searchArea, app_id, app_code]

        return restAPIURLComponents.url
    }

    /**
    Creates an array of annotations given the search result from the places demo API
    - Parameter searchResult: dictionary form of the JSON result of the query to
    the places demo API

    - Parameter searchString: place identifier searched
    - Parameter coordinate: ser location searched

    - Returns: Array of annotations if data of surrounding places is extracted from
    the searchResult, nil otherwise
    */
    func annotationsForSearchResult(searchResult : [String: AnyObject]?, searchString : String, coordinate: CLLocationCoordinate2D) -> [MKAnnotation]? {
        //parse search result into an array of Place objects
        if let places = self.dbhelper?.placesForSearchResult(searchResult: searchResult, searchString:searchString, coordinate: coordinate)
        {
          //create annotations from array of place objects
            return self.annotationsForPlaces(places: places)
        }
        return nil
    }

    /**
    Creates an array of annotations given an array of Place objects

    - Parameter places: array of place objects

    - Returns: Array of annotations
    */
    func annotationsForPlaces(places : [Place]?) -> [MKAnnotation]?{
        guard let placeArray = places else {
            return nil
        }

        var annotations : [MKAnnotation] = []
        for place in placeArray {

            var unit = "m"
            var unitDivisor = 1.0
            if place.distance >= 1000.0
            {
                unit = "km"
                unitDivisor = 1000.0
            }
            let distanceString = NSString(format:"%g", place.distance/unitDivisor)

            let annotation = MKPointAnnotation()
            annotation.coordinate.latitude = place.latitude
            annotation.coordinate.longitude = place.longitude
            annotation.title =  (place.title ?? "") + ", "  + (place.vicinity ?? "")
            annotation.subtitle = "\(distanceString) \(unit) away"

            annotations.append(annotation)
        }

        return annotations
    }

    /**
    Shows an alert for the given alert message

    - Parameter message: message to display in alert
    */
    func showAlert(message: String){
        let appName : String? = Bundle.main.infoDictionary?["CFBundleDisplayName"]  as? String ?? ""

        let alert  = UIAlertController(title:appName, message:message, preferredStyle: UIAlertController.Style.alert)

        let action = UIAlertAction(title: "OK", style: UIAlertAction.Style.cancel) { (alertaction) in

        }

        alert.addAction(action)

        self.present(alert, animated: true) {
            //Do nothing
        }
    }

    /**
    Sets the state of the UI when a search has ended.
    * Stop searching searchingIndicator
    * enable refreshButton
    * enable user interaction for mapView and searchBar
    */
    func didEndSearch()
    {
        self.searchingIndicator.stopAnimating()
        self.refreshButton.isEnabled = true
        self.mapView.isUserInteractionEnabled = true
        self.searchBar.isUserInteractionEnabled = true
    }
}
